src/libstore/derivations.cc

```diff
@@ -272,17 +272,19 @@ Derivation parseDerivation(const Store & store, std::string && s, std::string_vi
 
 static void printString(string & res, std::string_view s)
 {
-    char buf[s.size() * 2 + 2];
-    char * p = buf;
-    *p++ = '"';
-    for (auto c : s)
-        if (c == '\"' || c == '\\') { *p++ = '\\'; *p++ = c; }
-        else if (c == '\n') { *p++ = '\\'; *p++ = 'n'; }
-        else if (c == '\r') { *p++ = '\\'; *p++ = 'r'; }
-        else if (c == '\t') { *p++ = '\\'; *p++ = 't'; }
-        else *p++ = c;
-    *p++ = '"';
-    res.append(buf, p - buf);
+    size_t bufSize = s.size() * 2 + 2;
+    withBuffer(bufSize, [&](char *buf) {
```

Comment:
Maybe we can use `boost::container::small_vector<char, N>` instead? See c9fc975259e27220caeb4291f3dff453e65f1965 for an example.

Reply:
This causes a significant stack allocation even when we use the heap. Should we sacrifice that?

Alternatively I could change `withBuffer` to a RAII type instead of a higher order function.

Reply:
Stack allocations are essentially free (in a non-recursive function) so I don't think that's a problem.

Reply:
Update: can't convert `withBuffer` to RAII without `alloca`, because the number of elements would need to be static. I've tried a couple of things but couldn't make it work, and it would make sense that structs and stack frames are different, even if they look similar.

@pennae What do you think? Should we sacrifice some stack space to use boost `small_vector`?

Reply:
What's the concern with stack space? A stack allocation is at most one machine instruction.

Reply:
Oh, didn't see @edolstra's message; didn't mean to ignore him.

> Stack allocations are essentially free

I figured there's the cost of ~1 extra page in page tables or whatnot, as the `res.append` stack frame shifts, but I don't know if that's significant.

Reply:
@roberth even if we do "waste" stack space with a `small_vector`, we'll still waste less than before with your current stack buffer size. page faults are amortized, and we're not in the territory where we have to worry a lot about TLB misses. if it makes the code more readable we'd say it's well worth it :)

Reply:
@edolstra Done. Could you review?